disk-buffering/src/main/java/io/opentelemetry/contrib/disk/buffering/internal/storage/FolderManager.java

```diff
@@ -0,0 +1,145 @@
+import java.io.File;
+import java.io.IOException;
+import java.util.Objects;
+import javax.annotation.Nullable;
+
+public final class FolderManager {
+  private final File folder;
+  private final TimeProvider timeProvider;
+  private final StorageConfiguration configuration;
+  @Nullable private ReadableFile currentReadableFile;
+  @Nullable private WritableFile currentWritableFile;
+
+  public FolderManager(File folder, StorageConfiguration configuration) {
+    this(folder, configuration, TimeProvider.get());
+  }
+
+  public FolderManager(File folder, StorageConfiguration configuration, TimeProvider timeProvider) {
+    this.folder = folder;
+    this.configuration = configuration;
+    this.timeProvider = timeProvider;
+  }
+
+  @Nullable
+  public synchronized ReadableFile getReadableFile() throws IOException {
+    currentReadableFile = null;
+    File readableFile = findReadableFile();
+    if (readableFile != null) {
+      currentReadableFile =
+          new ReadableFile(
+              readableFile, Long.parseLong(readableFile.getName()), timeProvider, configuration);
+      return currentReadableFile;
+    }
+    return null;
+  }
+
+  public synchronized WritableFile createWritableFile() throws IOException {
+    long systemCurrentTimeMillis = timeProvider.getSystemCurrentTimeMillis();
+    File[] existingFiles = folder.listFiles();
+    if (existingFiles != null) {
+      if (purgeExpiredFilesIfAny(existingFiles, systemCurrentTimeMillis) == 0) {
+        removeOldestFileIfSpaceIsNeeded(existingFiles);
+      }
+    }
+    File file = new File(folder, String.valueOf(systemCurrentTimeMillis));
+    currentWritableFile =
+        new WritableFile(file, systemCurrentTimeMillis, configuration, timeProvider);
+    return currentWritableFile;
+  }
+
+  @Nullable
+  private File findReadableFile() throws IOException {
+    long currentTime = timeProvider.getSystemCurrentTimeMillis();
+    File[] existingFiles = folder.listFiles();
+    File oldestFileAvailable = null;
+    long oldestFileCreationTimeMillis = 0;
+    if (existingFiles != null) {
+      for (File existingFile : existingFiles) {
+        long existingFileCreationTimeMillis = Long.parseLong(existingFile.getName());
+        if (isReadyToBeRead(currentTime, existingFileCreationTimeMillis)
+            && !hasExpiredForReading(currentTime, existingFileCreationTimeMillis)) {
+          if (oldestFileAvailable == null
+              || existingFileCreationTimeMillis < oldestFileCreationTimeMillis) {
+            oldestFileCreationTimeMillis = existingFileCreationTimeMillis;
+            oldestFileAvailable = existingFile;
+          }
+        }
+      }
+    }
+    // Checking if the oldest available file is currently the writable file.
+    if (oldestFileAvailable != null
+        && currentWritableFile != null
+        && oldestFileAvailable.equals(currentWritableFile.file)) {
+      currentWritableFile.close();
```

Comment:
how can this condition be true? I thought the reader and writer live in different processes?

Reply:
It's probably a bit of an edge-case, though since the reading and writing processes can be done in parallel, there's a case in which a file is open for writing new data for too long (and stopped receiving new data), and at the same time a reader in a separate thread would want to start reading from the oldest available file, and there might only be a single file in the folder which happens to be the one open for writing because it's been idle for a while so the checking for closing it has not run yet. So in that case this validation makes sure to close said file before opening it up for reading to avoid IO errors. This can happen because both processes (write/read) can be done in parallel, now, in most cases they will be done in separate files, but it's not impossible that they might clash into the same one if there are no other files available to be read and the writing process is idle to the point that its writing timeout expires.

Reply:
> and at the same time a reader in a separate thread ...

This is the critical part I'm trying to understand. Is the assumption that reader and writer
- always use the same JVM?
- always use the same instance of the storage object?

Reply:
That's correct. The storage object is not exposed from the signal exporter, and there should be a single exporter per signal which is used for both read/write operations. The writing process happens automatically as new signals are provided from the signal processor, whereas the reading process is triggered (using the same signal disk exporter) elsewhere in the consumer of the lib.